<a href="https://colab.research.google.com/github/BharathReddyRamasani/AI_DS/blob/main/gpt_model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Data Preparation: Load Dataset

In [ ]:
texts = [
    "The cat chased the mouse.",
    "The dog barked loudly.",
    "Machine learning is powerful.",
    "Deep learning uses neural networks."
]

print("Original texts:")
for text in texts:
    print(text)

Original texts:
The cat chased the mouse.
The dog barked loudly.
Machine learning is powerful.
Deep learning uses neural networks.


## Data Preparation: Convert to Lowercase

In [ ]:
texts_lower = [text.lower() for text in texts]

print("Texts after converting to lowercase:")
for text in texts_lower:
    print(text)

Texts after converting to lowercase:
the cat chased the mouse.
the dog barked loudly.
machine learning is powerful.
deep learning uses neural networks.


## Data Preparation: Remove Unwanted Symbols

In [ ]:
import re

def remove_symbols(text):
    return re.sub(r'[^a-z0-9\s]', '', text)

texts_cleaned = [remove_symbols(text) for text in texts_lower]

print("Texts after removing unwanted symbols:")
for text in texts_cleaned:
    print(text)

Texts after removing unwanted symbols:
the cat chased the mouse
the dog barked loudly
machine learning is powerful
deep learning uses neural networks


## Data Preparation: Tokenize Text

In [ ]:
texts_tokenized = [text.split() for text in texts_cleaned]

print("Texts after tokenization:")
for tokens in texts_tokenized:
    print(tokens)

Texts after tokenization:
['the', 'cat', 'chased', 'the', 'mouse']
['the', 'dog', 'barked', 'loudly']
['machine', 'learning', 'is', 'powerful']
['deep', 'learning', 'uses', 'neural', 'networks']


## Data Preparation: Build Vocabulary

In [ ]:
vocab = set()
for tokens in texts_tokenized:
    for token in tokens:
        vocab.add(token)

vocab = sorted(list(vocab))

word_to_idx = {word: i for i, word in enumerate(vocab)}
idx_to_word = {i: word for i, word in enumerate(vocab)}

print("Vocabulary:", vocab)
print("Word to index mapping:", word_to_idx)
print("Index to word mapping:", idx_to_word)

Vocabulary: ['barked', 'cat', 'chased', 'deep', 'dog', 'is', 'learning', 'loudly', 'machine', 'mouse', 'networks', 'neural', 'powerful', 'the', 'uses']
Word to index mapping: {'barked': 0, 'cat': 1, 'chased': 2, 'deep': 3, 'dog': 4, 'is': 5, 'learning': 6, 'loudly': 7, 'machine': 8, 'mouse': 9, 'networks': 10, 'neural': 11, 'powerful': 12, 'the': 13, 'uses': 14}
Index to word mapping: {0: 'barked', 1: 'cat', 2: 'chased', 3: 'deep', 4: 'dog', 5: 'is', 6: 'learning', 7: 'loudly', 8: 'machine', 9: 'mouse', 10: 'networks', 11: 'neural', 12: 'powerful', 13: 'the', 14: 'uses'}


## Data Preparation: Create Input-Output Sequences

In [ ]:
input_sequences = []
target_words = []

for tokens in texts_tokenized:
    for i in range(1, len(tokens)): # Start from the second word to have a target
        input_sequence = tokens[:i] # All words up to the current one
        target_word = tokens[i] # The current word is the target

        # Convert input sequence to indices
        input_sequence_indices = [word_to_idx[word] for word in input_sequence]

        input_sequences.append(input_sequence_indices)
        target_words.append(word_to_idx[target_word])

# Display some examples
print("Example Input-Output Sequences:")
for i in range(min(5, len(input_sequences))):
    input_str = ' '.join([idx_to_word[idx] for idx in input_sequences[i]])
    target_str = idx_to_word[target_words[i]]
    print(f"Input: '{input_str}' -> Target: '{target_str}'")

Example Input-Output Sequences:
Input: 'the' -> Target: 'cat'
Input: 'the cat' -> Target: 'chased'
Input: 'the cat chased' -> Target: 'the'
Input: 'the cat chased the' -> Target: 'mouse'
Input: 'the' -> Target: 'dog'


## Task 2: Token Embeddings

### Initialize Token Embeddings

In [ ]:
import numpy as np

vocab_size = len(vocab)
embedding_dim = 8 # Changed to be divisible by num_heads (4)

# Initialize random embeddings for each word in the vocabulary
# In a real scenario, these would be learned during training or loaded from pre-trained models
embedding_matrix = np.random.rand(vocab_size, embedding_dim)

print(f"Embedding matrix shape: {embedding_matrix.shape}")
print("First 5 embeddings (for demonstration):")
for i in range(min(5, vocab_size)):
    word = idx_to_word[i]
    print(f"Word: '{word}' -> Embedding: {embedding_matrix[i][:5]}...")

Embedding matrix shape: (15, 8)
First 5 embeddings (for demonstration):
Word: 'barked' -> Embedding: [0.77474169 0.62822255 0.72789989 0.13485514 0.71205607]...
Word: 'cat' -> Embedding: [0.56354001 0.16132812 0.93888779 0.98179504 0.98011051]...
Word: 'chased' -> Embedding: [0.08891426 0.88396581 0.02987746 0.50671262 0.59879296]...
Word: 'deep' -> Embedding: [0.39218939 0.18501551 0.19418392 0.85269004 0.36870748]...
Word: 'dog' -> Embedding: [0.06923151 0.89007177 0.10173595 0.81434414 0.52369973]...


### Convert Token IDs to Embeddings

*   List item
*   List item



In [ ]:
embedded_input_sequences = []
embedded_target_words = []

for seq_indices in input_sequences:
    embedded_seq = [embedding_matrix[idx] for idx in seq_indices]
    embedded_input_sequences.append(embedded_seq)

for target_idx in target_words:
    embedded_target_words.append(embedding_matrix[target_idx])

print("Example of an embedded input sequence (first sequence, first token):")
if embedded_input_sequences:
    print(f"Original token ID: {input_sequences[0][0]}")
    print(f"Embedded vector: {embedded_input_sequences[0][0]}")

print("\nExample of an embedded target word (first target):")
if embedded_target_words:
    print(f"Original token ID: {target_words[0]}")
    print(f"Embedded vector: {embedded_target_words[0]}")

Example of an embedded input sequence (first sequence, first token):
Original token ID: 13
Embedded vector: [0.42746313 0.28731316 0.99180099 0.80588054 0.64220593 0.69621844
 0.31477195 0.64835509 0.04194683 0.88919126]

Example of an embedded target word (first target):
Original token ID: 1
Embedded vector: [0.88104509 0.7910065  0.98558783 0.7720128  0.43563495 0.5570996
 0.38442532 0.19313799 0.65515792 0.95485852]


## Task 3: Positional Encoding

### Add Positional Information to Input Sequences

Positional encoding adds information about the relative or absolute position of tokens in the sequence. Since the token embeddings we generated don't inherently contain order information, positional encoding is used to inject this into the model. For this example, we'll use a simple integer-based positional encoding, where each token's position in its sequence is added to its representation.

In [ ]:
# For simplicity, we'll create a new list of sequences with positional information.
# In a real model, this would often be added to the embeddings themselves.

input_sequences_with_pos = []

for seq_idx_list in input_sequences:
    seq_with_pos = []
    for i, token_idx in enumerate(seq_idx_list):
        # For demonstration, we'll store a tuple: (token_id, position)
        # In a deep learning model, this positional info might be added to the embedding vector.
        seq_with_pos.append((token_idx, i)) # i is the position
    input_sequences_with_pos.append(seq_with_pos)


print("Example Input Sequences with Positional Information:")
for i in range(min(5, len(input_sequences_with_pos))):
    original_tokens = [idx_to_word[item[0]] for item in input_sequences_with_pos[i]]
    positions = [item[1] for item in input_sequences_with_pos[i]]
    print(f"Sequence {i+1}: {original_tokens}")
    print(f"Positions   : {positions}")
    print(f"Tokens with positions: {input_sequences_with_pos[i]}")
    print("\n")

Example Input Sequences with Positional Information:
Sequence 1: ['the']
Positions   : [0]
Tokens with positions: [(13, 0)]


Sequence 2: ['the', 'cat']
Positions   : [0, 1]
Tokens with positions: [(13, 0), (1, 1)]


Sequence 3: ['the', 'cat', 'chased']
Positions   : [0, 1, 2]
Tokens with positions: [(13, 0), (1, 1), (2, 2)]


Sequence 4: ['the', 'cat', 'chased', 'the']
Positions   : [0, 1, 2, 3]
Tokens with positions: [(13, 0), (1, 1), (2, 2), (13, 3)]


Sequence 5: ['the']
Positions   : [0]
Tokens with positions: [(13, 0)]




## Task 4: Masked Self Attention

### Objective: Implement Causal Attention

Causal (or masked) self-attention is a core component of autoregressive models like GPT. It ensures that when generating a token at a certain position, the model can only attend to previous tokens in the sequence, not future ones. This prevents the model from 'cheating' by looking ahead at the answer.

Here, we'll demonstrate the creation of an attention mask, the calculation of attention scores, the application of the mask, and the derivation of attention weights for a sample input sequence.

### Select a Sample Sequence for Demonstration

We'll pick an input sequence with a few tokens to clearly illustrate the masking effect.

In [ ]:
# Using the 4th input sequence for demonstration: 'the cat chased the'
sample_sequence_idx = 3 # Corresponds to ['the', 'cat', 'chased', 'the'] in terms of actual words
sample_token_ids = input_sequences[sample_sequence_idx]

# Convert token IDs back to words for better readability
sample_words = [idx_to_word[idx] for idx in sample_token_ids]

# IMPORTANT: Re-generate sample_embedded_input using the current embedding_matrix to ensure correct dimensions (8).
# This addresses the inconsistency where `sample_embedded_input` or `embedded_input_sequences`
# might have contained stale data from previous runs when `embedding_dim` was potentially 10.
sample_embedded_input = [embedding_matrix[idx] for idx in sample_token_ids]

seq_len = len(sample_embedded_input)
# `embedding_dim` (which is 8 from cell c44fe1a5) is used here. Removed redundant `embedding_dim = embedding_dim`.

print(f"Selected sequence: {sample_words}")
print(f"Sequence length: {seq_len}")
print(f"Embedding dimension: {embedding_dim}")
print(f"Shape of embedded input: ({seq_len}, {embedding_dim})")

Selected sequence: ['the', 'cat', 'chased', 'the']
Sequence length: 4
Embedding dimension: 8
Shape of embedded input: (4, 8)


### 1. Create Causal Attention Mask

The mask is a square matrix where positions corresponding to future tokens are set to a very small number (e.g., negative infinity) so they become zero after softmax. Allowed positions are typically 0.

In [ ]:
causal_mask = np.full((seq_len, seq_len), -np.inf) # Initialize with negative infinity

# Fill the lower triangular part with zeros (allowing attention to previous and current tokens)
for i in range(seq_len):
    for j in range(i + 1):
        causal_mask[i, j] = 0

print("Causal Attention Mask:")
print(causal_mask)
print("\nThis mask shows that for each position (row), only previous or current positions (columns) are accessible (0), while future positions are masked out (-inf).")

Causal Attention Mask:
[[  0. -inf -inf -inf]
 [  0.   0. -inf -inf]
 [  0.   0.   0. -inf]
 [  0.   0.   0.   0.]]

This mask shows that for each position (row), only previous or current positions (columns) are accessible (0), while future positions are masked out (-inf).


### 2. Calculate Attention Scores (Dot Product)

For self-attention, the Query, Key, and Value matrices are typically derived from the same input. For this demonstration, we'll use the `sample_embedded_input` directly as our Q, K, and V (though in a real model, linear transformations would be applied to get Q, K, V). The attention scores are computed by a dot product between Query and Key.

In [ ]:
# For simplicity, Q = K = V = sample_embedded_input
Q = np.array(sample_embedded_input)
K = np.array(sample_embedded_input)

# Calculate raw attention scores (Query * Key_transpose)
# (Q @ K.T)
attention_scores = np.matmul(Q, K.T)

# Scale the scores (common practice in Transformers)
attention_scores = attention_scores / np.sqrt(embedding_dim)

print("Raw Attention Scores (Q * K.T / sqrt(d_k)):")
print(attention_scores)
print("\nEach row represents a query token, and each column represents a key token. The value indicates their similarity.")

Raw Attention Scores (Q * K.T / sqrt(d_k)):
[[1.29887429 1.26299164 1.01279637 1.29887429]
 [1.26299164 1.5797248  1.19700521 1.26299164]
 [1.01279637 1.19700521 1.37218494 1.01279637]
 [1.29887429 1.26299164 1.01279637 1.29887429]]

Each row represents a query token, and each column represents a key token. The value indicates their similarity.


### 3. Apply Mask and Calculate Attention Weights

Now, we apply the causal mask to the attention scores. This sets the scores for future tokens to negative infinity. Then, a softmax function is applied to convert these scores into probability distributions (attention weights).

In [ ]:
def softmax(x):
    e_x = np.exp(x - np.max(x, axis=-1, keepdims=True)) # Subtract max for numerical stability
    return e_x / e_x.sum(axis=-1, keepdims=True)

# Apply the mask to the attention scores
masked_attention_scores = attention_scores + causal_mask

print("Masked Attention Scores:")
print(masked_attention_scores)

# Calculate attention weights using softmax
attention_weights = softmax(masked_attention_scores)

print("\nAttention Weights (after softmax):")
print(attention_weights)

print("\nDeliverables:")
print("Attention Mask (causal_mask):\n", causal_mask)
print("\nAttention Scores (attention_scores):\n", attention_scores)
print("\nAttention Weights (attention_weights):\n", attention_weights)
print("\nNotice that in the attention weights, future tokens now have 0 probability, ensuring causal attention.")

Masked Attention Scores:
[[1.29887429       -inf       -inf       -inf]
 [1.26299164 1.5797248        -inf       -inf]
 [1.01279637 1.19700521 1.37218494       -inf]
 [1.29887429 1.26299164 1.01279637 1.29887429]]

Attention Weights (after softmax):
[[1.         0.         0.         0.        ]
 [0.42147211 0.57852789 0.         0.        ]
 [0.27512434 0.33077289 0.39410277 0.        ]
 [0.26910964 0.25962447 0.20215626 0.26910964]]

Deliverables:
Attention Mask (causal_mask):
 [[  0. -inf -inf -inf]
 [  0.   0. -inf -inf]
 [  0.   0.   0. -inf]
 [  0.   0.   0.   0.]]

Attention Scores (attention_scores):
 [[1.29887429 1.26299164 1.01279637 1.29887429]
 [1.26299164 1.5797248  1.19700521 1.26299164]
 [1.01279637 1.19700521 1.37218494 1.01279637]
 [1.29887429 1.26299164 1.01279637 1.29887429]]

Attention Weights (attention_weights):
 [[1.         0.         0.         0.        ]
 [0.42147211 0.57852789 0.         0.        ]
 [0.27512434 0.33077289 0.39410277 0.        ]
 [0.26910964

## Task 5: Multi Head Attention

### Objective: Implement 4 Attention Heads and Analyze Their Roles

Multi-head attention allows a model to jointly attend to information from different representation subspaces at different positions. Instead of performing a single attention function with one set of queries, keys, and values, the multi-head mechanism linearly projects the queries, keys, and values `h` times with different, learned linear projections to `h` different sets of projected queries, keys, and values. Then, for each of these projected versions, the attention function is performed in parallel, yielding `h` output values. These are concatenated and again linearly projected to produce the final values.

For this demonstration, we will simplify by splitting the existing embeddings across heads and then conceptually describe what each head *might* focus on.

### Set up Multi-Head Parameters

In [ ]:
num_heads = 4
head_dim = embedding_dim // num_heads # Each head gets a portion of the embedding dimension

# Ensure embedding_dim is divisible by num_heads
if embedding_dim % num_heads != 0:
    raise ValueError("embedding_dim must be divisible by num_heads")

print(f"Number of attention heads: {num_heads}")
print(f"Dimension per head: {head_dim}")

Number of attention heads: 4
Dimension per head: 2


### Implement Multi-Head Causal Attention

We will take our `sample_embedded_input` and conceptually split it into `num_heads` parts to simulate separate Q, K, V projections for each head. Then, we apply the causal attention mask and calculate weights for each head.

In [ ]:
all_head_outputs = []

print("\n--- Multi-Head Attention Processing ---")
for i in range(num_heads):
    print(f"\nProcessing Head {i+1}:")

    # Simulate linear projections by slicing the embedding dimension for each head
    # In a real model, Q, K, V would be distinct linear transformations of input
    head_start_idx = i * head_dim
    head_end_idx = (i + 1) * head_dim

    Q_head = np.array([vec[head_start_idx:head_end_idx] for vec in sample_embedded_input])
    K_head = np.array([vec[head_start_idx:head_end_idx] for vec in sample_embedded_input])
    V_head = np.array([vec[head_start_idx:head_end_idx] for vec in sample_embedded_input])

    # Calculate raw attention scores for this head
    attention_scores_head = np.matmul(Q_head, K_head.T)
    attention_scores_head = attention_scores_head / np.sqrt(head_dim)

    # Apply causal mask
    masked_attention_scores_head = attention_scores_head + causal_mask

    # Calculate attention weights for this head
    attention_weights_head = softmax(masked_attention_scores_head)

    print(f"Head {i+1} Attention Weights:\n{attention_weights_head}")

    # Compute output for this head (Attention Weights @ V_head)
    output_head = np.matmul(attention_weights_head, V_head)
    all_head_outputs.append(output_head)

# Concatenate all head outputs
multi_head_output = np.concatenate(all_head_outputs, axis=-1)

print("\nConcatenated Multi-Head Output Shape:", multi_head_output.shape)
print("First 5 elements of Multi-Head Output (for token 1):\n", multi_head_output[0, :5])


--- Multi-Head Attention Processing ---

Processing Head 1:
Head 1 Attention Weights:
[[1.         0.         0.         0.        ]
 [0.5071932  0.4928068  0.         0.        ]
 [0.3953635  0.23948088 0.36515562 0.        ]
 [0.2898245  0.17215102 0.24819999 0.2898245 ]]

Processing Head 2:
Head 2 Attention Weights:
[[1.         0.         0.         0.        ]
 [0.31874885 0.68125115 0.         0.        ]
 [0.30353138 0.3811184  0.31535022 0.        ]
 [0.23314196 0.31970903 0.21400706 0.23314196]]

Processing Head 3:
Head 3 Attention Weights:
[[1.         0.         0.         0.        ]
 [0.47324216 0.52675784 0.         0.        ]
 [0.302555   0.32378385 0.37366115 0.        ]
 [0.24871399 0.27223901 0.23033302 0.24871399]]

Processing Head 4:
Head 4 Attention Weights:
[[1.         0.         0.         0.        ]
 [0.33730964 0.66269036 0.         0.        ]
 [0.30753098 0.38692146 0.30554756 0.        ]
 [0.2365523  0.29244497 0.23445043 0.2365523 ]]

Concatenated Multi

### Conceptual Analysis of Attention Heads

In a fully trained Transformer, each attention head often specializes in capturing different types of relationships. While our current model is not trained, we can conceptualize what each head *might* learn:

*   **Head 1 (Grammar/Syntax)**: Might focus on relationships between words that follow grammatical rules, such as subject-verb agreement or identifying phrases. For example, it could learn that an adjective typically precedes a noun.
*   **Head 2 (Context/Local Dependencies)**: Could attend to words that are semantically close or provide immediate context for the current word. For instance, in 'the cat chased the mouse', this head might connect 'cat' to 'chased' and 'chased' to 'mouse'.
*   **Head 3 (Long Dependencies/Coreference)**: Might look for connections between words that are far apart in the sequence but are semantically related (e.g., coreference resolution). In a longer text, this head could link a pronoun to its antecedent.
*   **Head 4 (Semantics/Entity Recognition)**: Could focus on identifying types of entities (e.g., 'cat' and 'mouse' are animals) or broader semantic categories.

The final multi-head output combines these diverse perspectives, providing a richer representation for each token.

## Task 6: GPT Decoder Block

### Add & Normalize Layer

This layer combines a residual connection with the output of a sub-layer (like attention or feed-forward) and then applies layer normalization. Layer normalization helps stabilize training and improve performance.

In [ ]:
def add_and_normalize(x, sublayer_output, epsilon=1e-6):
    """
    Applies residual connection followed by layer normalization.
    x: The input to the sub-layer (for residual connection).
    sublayer_output: The output of the sub-layer.
    epsilon: Small value to prevent division by zero in normalization.
    """
    # Add residual connection
    added_output = x + sublayer_output

    # Layer Normalization
    mean = np.mean(added_output, axis=-1, keepdims=True)
    variance = np.var(added_output, axis=-1, keepdims=True)
    normalized_output = (added_output - mean) / np.sqrt(variance + epsilon)

    # In a real model, learnable gamma and beta parameters would be applied here.
    # For this conceptual implementation, we omit them.
    return normalized_output

print("Defined add_and_normalize function.")

Defined add_and_normalize function.


### Feed Forward Network

The Feed Forward Network (FFN) consists of two linear transformations with a non-linear activation function (typically ReLU) in between. It processes each position independently and identically.

In [ ]:
def feed_forward_network(x, d_model, d_ff):
    """
    Applies a simple feed-forward network to the input.
    x: Input tensor.
    d_model: Dimension of input/output.
    d_ff: Dimension of the inner layer of the FFN.
    """
    # Simulate linear transformations (weights and biases are omitted for simplicity)
    # First linear layer (d_model -> d_ff)
    # Assuming random weights for demonstration
    W1 = np.random.rand(d_model, d_ff) * 0.01 # Small weights for stability
    b1 = np.random.rand(d_ff) * 0.01
    linear1 = np.dot(x, W1) + b1

    # ReLU activation
    relu_output = np.maximum(0, linear1)

    # Second linear layer (d_ff -> d_model)
    W2 = np.random.rand(d_ff, d_model) * 0.01
    b2 = np.random.rand(d_model) * 0.01
    output = np.dot(relu_output, W2) + b2

    return output

print("Defined feed_forward_network function.")

Defined feed_forward_network function.


### GPT Decoder Block Implementation

Now, let's assemble the components into a full GPT Decoder Block. A typical decoder block in GPT consists of:

1.  Masked Multi-Head Attention
2.  Add & Normalize (residual connection + layer normalization)
3.  Feed Forward Network
4.  Add & Normalize (residual connection + layer normalization)


In [ ]:
def gpt_decoder_block(x, d_model, d_ff, num_heads, causal_mask):
    """
    Implements a single GPT Decoder Block, including internal Masked Multi-Head Attention.
    x: The input to the block (e.g., embedded input + positional encoding).
    d_model: The embedding dimension (also the output dimension of the block).
    d_ff: The inner dimension of the Feed Forward Network.
    num_heads: Number of attention heads for Multi-Head Attention.
    causal_mask: The causal mask to apply during attention.
    """

    # 1. Masked Multi-Head Attention
    head_dim = d_model // num_heads
    all_head_outputs = []

    # Simulate linear projections by slicing the embedding dimension for each head
    # In a real model, Q, K, V would be distinct linear transformations of input 'x'
    # For simplicity, we use 'x' as source for Q, K, V and slice it for heads.
    for i in range(num_heads):
        head_start_idx = i * head_dim
        head_end_idx = (i + 1) * head_dim

        Q_head = np.array([vec[head_start_idx:head_end_idx] for vec in x])
        K_head = np.array([vec[head_start_idx:head_end_idx] for vec in x])
        V_head = np.array([vec[head_start_idx:head_end_idx] for vec in x])

        # Calculate raw attention scores for this head
        attention_scores_head = np.matmul(Q_head, K_head.T)
        attention_scores_head = attention_scores_head / np.sqrt(head_dim)

        # Apply causal mask
        masked_attention_scores_head = attention_scores_head + causal_mask

        # Calculate attention weights for this head
        attention_weights_head = softmax(masked_attention_scores_head)

        # Compute output for this head (Attention Weights @ V_head)
        output_head = np.matmul(attention_weights_head, V_head)
        all_head_outputs.append(output_head)

    multi_head_output = np.concatenate(all_head_outputs, axis=-1)

    # 2. Add & Normalize after Masked Multi-Head Attention
    # The input `x` serves as the residual connection for the attention sub-layer.
    norm1_output = add_and_normalize(x, multi_head_output)

    # 3. Feed Forward Network
    ffn_output = feed_forward_network(norm1_output, d_model, d_ff)

    # 4. Add & Normalize after Feed Forward Network
    # `norm1_output` serves as the residual connection for the FFN sub-layer.
    decoder_block_output = add_and_normalize(norm1_output, ffn_output)

    return decoder_block_output

print("Modified gpt_decoder_block function to include Multi-Head Attention internally.")

Modified gpt_decoder_block function to include Multi-Head Attention internally.


### Demonstration of the GPT Decoder Block

Let's apply our `gpt_decoder_block` using the `multi_head_output` from the previous task. We'll use the `sample_embedded_input` as the initial input `x` for the decoder block, as it represents the token embeddings with which the residual connection starts.

In [ ]:
d_model = embedding_dim # This is our embedding_dim (8)
d_ff = 2 * d_model # A common choice for the inner dimension of FFN

print(f"d_model (embedding_dim): {d_model}")
print(f"d_ff (Feed Forward inner dimension): {d_ff}")

# The input `x` to the decoder block is typically the output of the previous block
# or the embedded input + positional encoding for the first block.
# Here, we'll use `sample_embedded_input` as the starting point for `x`.
initial_block_input = np.array(sample_embedded_input) # Ensure it's a numpy array

# Recalculate causal mask for the sample sequence
seq_len = initial_block_input.shape[0]
causal_mask_demo = np.full((seq_len, seq_len), -np.inf)
for i in range(seq_len):
    for j in range(i + 1):
        causal_mask_demo[i, j] = 0

print(f"Number of heads for demo: {num_heads}")

# Run the GPT Decoder Block with the modified signature
decoder_output = gpt_decoder_block(initial_block_input, d_model, d_ff, num_heads, causal_mask_demo)

print("\nGPT Decoder Block Output Shape:", decoder_output.shape)
print("First 5 elements of GPT Decoder Block Output (for token 1):\n", decoder_output[0, :5])
print("\nThis demonstrates a single pass through a GPT-like decoder block.")

d_model (embedding_dim): 8
d_ff (Feed Forward inner dimension): 16
Number of heads for demo: 4

GPT Decoder Block Output Shape: (4, 8)
First 5 elements of GPT Decoder Block Output (for token 1):
 [-0.28494457  1.74110694 -0.19014109 -0.36499823  1.20292711]

This demonstrates a single pass through a GPT-like decoder block.


## Task 7: GPT Model

Now, let's assemble all the components we've built into a complete, simplified GPT model architecture:

1.  **Input Tokens**
2.  **Embedding Layer**
3.  **Position Encoding**
4.  **Decoder Block (repeated N times)**
5.  **Linear Layer (for vocabulary projection)**
6.  **Softmax (for probability distribution over vocabulary)**

### Positional Encoding Function

We'll define a simple positional encoding function that adds a numerical representation of a token's position to its embedding. This is crucial as attention mechanisms are permutation-invariant, meaning they don't inherently understand word order without this explicit positional information.

In [ ]:
def create_positional_encoding(seq_len, d_model):
    """
    Generates a simplified additive positional encoding.
    In a real Transformer, this would typically use sin/cos functions.
    """
    position = np.arange(seq_len)[:, np.newaxis] # (seq_len, 1)

    # Scale position by a small factor and tile it across d_model dimensions
    # This is a very simplified version for conceptual understanding.
    if seq_len > 1:
        pos_encoding = np.tile(position / (seq_len - 1), (1, d_model))
    else:
        pos_encoding = np.zeros((seq_len, d_model)) # Handle single token sequence

    return pos_encoding

print("Defined create_positional_encoding function.")

Defined create_positional_encoding function.


### GPT Model Implementation

This function will integrate all the modules: embedding, positional encoding, stacked decoder blocks, a linear output layer, and softmax to produce token probabilities.

In [ ]:
def gpt_model(token_ids_sequence, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks):
    """
    Implements a simplified GPT Model architecture.
    token_ids_sequence: List of token IDs (e.g., [13, 1, 2])
    embedding_matrix: Pre-initialized word embeddings (vocab_size, d_model)
    vocab_size: Size of the vocabulary
    d_model: Dimension of embeddings and decoder block outputs
    d_ff: Inner dimension of the Feed Forward Networks
    num_heads: Number of attention heads
    num_decoder_blocks: Number of stacked decoder blocks
    """
    seq_len = len(token_ids_sequence)

    # 1. Embedding Layer
    embedded_input = np.array([embedding_matrix[idx] for idx in token_ids_sequence])

    # 2. Positional Encoding
    pos_encoding = create_positional_encoding(seq_len, d_model)
    x = embedded_input + pos_encoding # Add positional information to embeddings

    # Create causal mask for the entire sequence (needed for each decoder block)
    causal_mask_model = np.full((seq_len, seq_len), -np.inf)
    for i in range(seq_len):
        for j in range(i + 1):
            causal_mask_model[i, j] = 0

    # 3. Stack Decoder Blocks
    for block_num in range(num_decoder_blocks):
        x = gpt_decoder_block(x, d_model, d_ff, num_heads, causal_mask_model)

    # 4. Linear Layer (Projection to vocabulary size)
    # Simulate a linear layer: output = x @ W_out + b_out
    # Initialize random weights and biases for demonstration
    W_out = np.random.rand(d_model, vocab_size) * 0.01
    b_out = np.random.rand(vocab_size) * 0.01
    logits = np.dot(x, W_out) + b_out

    # 5. Softmax to get probabilities
    probabilities = softmax(logits)

    return probabilities

print("Defined gpt_model function.")

Defined gpt_model function.


### Demonstration of the GPT Model

Let's run our full GPT model with a sample input sequence and observe its output probabilities for the next token at each position. We'll use the same input sequence `['the', 'cat', 'chased', 'the']`.

In [ ]:
# Parameters for the GPT model
num_decoder_blocks = 2 # As per the architecture sketch

print(f"GPT Model Parameters:")
print(f"  Vocabulary size: {vocab_size}")
print(f"  Embedding dimension (d_model): {d_model}")
print(f"  Feed Forward inner dimension (d_ff): {d_ff}")
print(f"  Number of attention heads: {num_heads}")
print(f"  Number of decoder blocks: {num_decoder_blocks}")

# Use the sample token IDs from previous tasks
sample_token_ids_model = sample_token_ids # [13, 1, 2, 13] for 'the cat chased the'

print(f"\nInput token IDs: {sample_token_ids_model}")
print(f"Input sequence words: {[idx_to_word[idx] for idx in sample_token_ids_model]}")

# Run the GPT model
output_probabilities = gpt_model(sample_token_ids_model, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks)

print("\nGPT Model Output Probabilities Shape:", output_probabilities.shape)

print("\nOutput Probabilities for each token in the sequence (top 3 predictions):")
for i, token_id in enumerate(sample_token_ids_model):
    print(f"--- Position {i} (Input: '{idx_to_word[token_id]}') ---")
    # Get top 3 predicted tokens and their probabilities
    top_indices = np.argsort(output_probabilities[i])[::-1][:3]
    for idx in top_indices:
        word = idx_to_word[idx]
        prob = output_probabilities[i, idx]
        print(f"  '{word}': {prob:.4f}")

print("\nThis demonstrates the full forward pass of a GPT-like model, predicting the next token at each step.")

GPT Model Parameters:
  Vocabulary size: 15
  Embedding dimension (d_model): 8
  Feed Forward inner dimension (d_ff): 16
  Number of attention heads: 4
  Number of decoder blocks: 2

Input token IDs: [13, 1, 2, 13]
Input sequence words: ['the', 'cat', 'chased', 'the']

GPT Model Output Probabilities Shape: (4, 15)

Output Probabilities for each token in the sequence (top 3 predictions):
--- Position 0 (Input: 'the') ---
  'is': 0.0673
  'deep': 0.0672
  'mouse': 0.0672
--- Position 1 (Input: 'cat') ---
  'dog': 0.0677
  'deep': 0.0673
  'networks': 0.0672
--- Position 2 (Input: 'chased') ---
  'uses': 0.0676
  'is': 0.0674
  'neural': 0.0671
--- Position 3 (Input: 'the') ---
  'is': 0.0673
  'uses': 0.0673
  'deep': 0.0672

This demonstrates the full forward pass of a GPT-like model, predicting the next token at each step.


## Task 7: GPT Model

Now, let's assemble all the components we've built into a complete, simplified GPT model architecture:

1.  **Input Tokens**
2.  **Embedding Layer**
3.  **Position Encoding**
4.  **Decoder Block (repeated N times)**
5.  **Linear Layer (for vocabulary projection)**
6.  **Softmax (for probability distribution over vocabulary)**

### Positional Encoding Function

We'll define a simple positional encoding function that adds a numerical representation of a token's position to its embedding. This is crucial as attention mechanisms are permutation-invariant, meaning they don't inherently understand word order without this explicit positional information.

In [ ]:
def create_positional_encoding(seq_len, d_model):
    """
    Generates a simplified additive positional encoding.
    In a real Transformer, this would typically use sin/cos functions.
    """
    position = np.arange(seq_len)[:, np.newaxis] # (seq_len, 1)

    # Scale position by a small factor and tile it across d_model dimensions
    # This is a very simplified version for conceptual understanding.
    if seq_len > 1:
        pos_encoding = np.tile(position / (seq_len - 1), (1, d_model))
    else:
        pos_encoding = np.zeros((seq_len, d_model)) # Handle single token sequence

    return pos_encoding

print("Defined create_positional_encoding function.")

Defined create_positional_encoding function.


### GPT Model Implementation

This function will integrate all the modules: embedding, positional encoding, stacked decoder blocks, a linear output layer, and softmax to produce token probabilities.

In [ ]:
def gpt_model(token_ids_sequence, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks):
    """
    Implements a simplified GPT Model architecture.
    token_ids_sequence: List of token IDs (e.g., [13, 1, 2])
    embedding_matrix: Pre-initialized word embeddings (vocab_size, d_model)
    vocab_size: Size of the vocabulary
    d_model: Dimension of embeddings and decoder block outputs
    d_ff: Inner dimension of the Feed Forward Networks
    num_heads: Number of attention heads
    num_decoder_blocks: Number of stacked decoder blocks
    """
    seq_len = len(token_ids_sequence)

    # 1. Embedding Layer
    embedded_input = np.array([embedding_matrix[idx] for idx in token_ids_sequence])

    # 2. Positional Encoding
    # Use the pre-computed sinusoidal positional_encoding_matrix
    # Ensure seq_len does not exceed max_seq_length for positional encoding
    if seq_len > max_seq_length:
        raise ValueError(f"Sequence length ({seq_len}) exceeds max_seq_length ({max_seq_length}) for positional encoding.")

    pos_encoding = positional_encoding_matrix[:seq_len, :]
    x = embedded_input + pos_encoding # Add positional information to embeddings

    # Create causal mask for the entire sequence (needed for each decoder block)
    causal_mask_model = np.full((seq_len, seq_len), -np.inf)
    for i in range(seq_len):
        for j in range(i + 1):
            causal_mask_model[i, j] = 0

    # 3. Stack Decoder Blocks
    for block_num in range(num_decoder_blocks):
        x = gpt_decoder_block(x, d_model, d_ff, num_heads, causal_mask_model)

    # 4. Linear Layer (Projection to vocabulary size)
    # Simulate a linear layer: output = x @ W_out + b_out
    # Initialize random weights and biases for demonstration
    W_out = np.random.rand(d_model, vocab_size) * 0.01
    b_out = np.random.rand(vocab_size) * 0.01
    logits = np.dot(x, W_out) + b_out

    # 5. Softmax to get probabilities
    probabilities = softmax(logits)

    return probabilities

print("Defined gpt_model function using sinusoidal positional encoding.")

Defined gpt_model function using sinusoidal positional encoding.


## Task: Text Generation for TinyStories

With our `gpt_model` now using the `gpt2` tokenizer's vocabulary and sinusoidal positional encoding, let's demonstrate its text generation capability. We'll reuse the `predict_next_token` and `generate_text` functions, but adapt them to use the components we've set up for the `TinyStories` project (e.g., `tokenizer`, `embedding_matrix_tinystories`, `new_vocab_size`).

In [ ]:
def predict_next_token_tinystories(input_text, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, tokenizer):
    """
    Predicts the next token given an input text using the TinyStories-adapted model.
    """
    # Tokenize the input text using the GPT2 tokenizer
    input_token_ids = tokenizer.encode(input_text, return_tensors='np')[0].tolist()

    if not input_token_ids:
        return "", 0.0 # No input, no prediction

    # Run the GPT model to get probabilities
    probabilities = gpt_model(input_token_ids, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks)

    # Get the probabilities for the last token in the sequence (which predicts the next token)
    last_token_probabilities = probabilities[-1]

    # Get the index of the token with the highest probability
    predicted_token_idx = np.argmax(last_token_probabilities)
    predicted_probability = last_token_probabilities[predicted_token_idx]

    # Decode the predicted token back to a word
    predicted_word = tokenizer.decode([predicted_token_idx])

    return predicted_word, predicted_probability


def generate_text_tinystories(prompt_text, max_length, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, tokenizer):
    """
    Generates text autoregressively from a given prompt using the TinyStories-adapted model.
    """
    # Tokenize the initial prompt
    generated_token_ids = tokenizer.encode(prompt_text, return_tensors='np')[0].tolist()

    for _ in range(max_length - len(generated_token_ids)): # Generate tokens until max_length is reached
        if not generated_token_ids: # Handle empty sequence if max_length is less than prompt length
            break

        # Predict the next token based on the current generated sequence
        probabilities = gpt_model(generated_token_ids, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks)

        # Get probabilities for the last token's prediction
        last_token_probabilities = probabilities[-1]

        # Select the token with the highest probability (greedy decoding)
        next_token_idx = np.argmax(last_token_probabilities)
        generated_token_ids.append(next_token_idx)

    # Convert token IDs back to words
    generated_text = tokenizer.decode(generated_token_ids)
    return generated_text

print("Defined predict_next_token_tinystories and generate_text_tinystories functions.")

Defined predict_next_token_tinystories and generate_text_tinystories functions.


In [ ]:
# Define global parameters for the GPT model in the TinyStories context
# d_model is already defined as 8 in cell ff45b1bd
d_ff = 2 * d_model # Inner dimension of the Feed Forward Network
num_decoder_blocks = 2 # Number of stacked decoder blocks
num_heads = 4 # Number of attention heads (already defined, but good to re-state for clarity)

print(f"Global Parameters for TinyStories GPT model:")
print(f"  d_model: {d_model}")
print(f"  d_ff: {d_ff}")
print(f"  num_decoder_blocks: {num_decoder_blocks}")
print(f"  num_heads: {num_heads}")

Global Parameters for TinyStories GPT model:
  d_model: 8
  d_ff: 16
  num_decoder_blocks: 2
  num_heads: 4


### Demonstration: Next Token Prediction for TinyStories

In [ ]:
# Example 1: "The little cat"
input_text_1 = "The little cat"

predicted_word_1, prob_1 = predict_next_token_tinystories(input_text_1, embedding_matrix_tinystories, new_vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, tokenizer)
print(f"Input: '{input_text_1}'\nPrediction: '{predicted_word_1}' (Probability: {prob_1:.4f})\n")

# Example 2: "Once upon a time, a"
input_text_2 = "Once upon a time, a"

predicted_word_2, prob_2 = predict_next_token_tinystories(input_text_2, embedding_matrix_tinystories, new_vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, tokenizer)
print(f"Input: '{input_text_2}'\nPrediction: '{predicted_word_2}' (Probability: {prob_2:.4f})\n")

print("Note: Predictions will still be random as the model is untrained.")

NameError: name 'gpt_decoder_block' is not defined

### Demonstration: Text Generation for TinyStories

In [ ]:
# Example: "The quick brown fox"
prompt_text_gen = "The quick brown fox"
max_gen_length = len(tokenizer.encode(prompt_text_gen)) + 10 # Generate 10 new tokens

generated_text = generate_text_tinystories(prompt_text_gen, max_gen_length, embedding_matrix_tinystories, new_vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, tokenizer)
print(f"Input Prompt: '{prompt_text_gen}'\nGeneration: {generated_text}")

print("\nNote: Text generation will be semantically incoherent because the model's weights are randomly initialized.")

NameError: name 'gpt_decoder_block' is not defined

## Task: Training the GPT Model (Conceptual)

Training a GPT model involves iteratively adjusting its weights so that it can accurately predict the next token in a sequence. For our conceptual NumPy model, we'll outline the key steps involved, rather than implementing a full, optimized training loop (which is typically handled by deep learning frameworks like PyTorch or TensorFlow).

The main components of training are:
1.  **Data Preparation for Training Batches:** Organizing the tokenized data into input sequences and corresponding target labels.
2.  **Loss Function:** A metric to quantify how well the model's predictions align with the actual next tokens.
3.  **Optimization (Conceptual):** Adjusting model weights based on the calculated loss (we'll only describe this conceptually).

Let's first prepare our data.

### 1. Data Preparation for Training Batches

Our `tokenized_dataset_subset` contains `input_ids`, `attention_mask`, and `labels`. For training, we primarily need the `input_ids` as input to our `gpt_model` and `labels` as the ground truth for loss calculation. We'll extract these from our subset.

In [ ]:
# Extract input_ids and labels from the tokenized dataset subset
training_input_ids = tokenized_dataset_subset['input_ids']
training_labels = tokenized_dataset_subset['labels']

print(f"Number of training samples: {len(training_input_ids)}")
print(f"Shape of first input_ids sequence: {np.array(training_input_ids[0]).shape}")
print(f"Shape of first labels sequence: {np.array(training_labels[0]).shape}")

# We'll use a small batch size for conceptual demonstration
batch_size = 4

print(f"\nConceptual Batch Size for Training: {batch_size}")
print("In a real training scenario, these would be fed to the model in batches.")

Number of training samples: 30000
Shape of first input_ids sequence: (128,)
Shape of first labels sequence: (128,)

Conceptual Batch Size for Training: 4
In a real training scenario, these would be fed to the model in batches.


### 2. Loss Function (Categorical Cross-Entropy)

For language modeling, the goal is to predict the next token. This is a multi-class classification problem where each token in the vocabulary is a possible class. **Categorical Cross-Entropy Loss** is typically used to measure the difference between the model's predicted probability distribution over the vocabulary and the true one-hot encoded distribution of the next token.

Since our `labels` are masked with -100 for padding, we need to ensure our loss function ignores these values.

In [ ]:
def categorical_cross_entropy_loss(predictions, targets, ignore_index=-100):
    """
    Calculates categorical cross-entropy loss, ignoring a specified index.

    predictions: (seq_len, vocab_size) array of predicted probabilities.
    targets: (seq_len,) array of true token indices.
    ignore_index: The index to ignore in the target sequence (e.g., padding).
    """
    # Ensure targets are valid indices
    valid_indices = targets != ignore_index
    if not np.any(valid_indices):
        return 0.0 # No valid tokens to compute loss for

    # Filter predictions and targets for valid tokens
    valid_predictions = predictions[valid_indices]
    valid_targets = targets[valid_indices]

    # Clip probabilities to avoid log(0)
    epsilon = 1e-9
    valid_predictions = np.clip(valid_predictions, epsilon, 1. - epsilon)

    # Convert valid_targets to one-hot encoding for comparison with predictions
    # Note: This is simplified; in full model, targets would be shifted for next-token pred.
    # For simplicity here, we assume predictions are for each position's target.
    one_hot_targets = np.zeros_like(valid_predictions)
    one_hot_targets[np.arange(len(valid_targets)), valid_targets] = 1

    loss = -np.sum(one_hot_targets * np.log(valid_predictions)) / len(valid_targets)
    return loss

print("Defined categorical_cross_entropy_loss function.")

Defined categorical_cross_entropy_loss function.


### 3. Conceptual Forward Pass for Training & Loss Calculation

In a training loop, for each batch:
1.  **Input Preparation:** Get a batch of `input_ids`.
2.  **Model Forward Pass:** Pass the `input_ids` through the `gpt_model` to get `output_probabilities` (shape `(batch_seq_len, vocab_size)` for each sequence).
3.  **Target Preparation:** Get the corresponding `labels` for the batch. These `labels` are typically shifted versions of the `input_ids` (e.g., if input is `[t1, t2, t3]`, target for prediction at `t1` is `t2`, at `t2` is `t3`, etc.). Our `tokenized_dataset_subset` already provides `labels` in this format with padding ignored.
4.  **Loss Calculation:** Compute the `categorical_cross_entropy_loss` between the `output_probabilities` and the `labels`.

In [ ]:
# Conceptual demonstration of loss calculation for a single sequence

# Take the first tokenized example from our subset
example_input_ids = training_input_ids[0]
example_labels = training_labels[0]

print(f"Example input_ids (first 10): {example_input_ids[:10]}...")
print(f"Example labels (first 10): {example_labels[:10]}...")

# Perform a forward pass with the untrained model
# Note: gpt_model expects a single sequence, so we can't directly batch here
# For simplicity, we'll calculate loss per sequence then average, or just for one.

# Filter out padding from labels for actual comparison if needed for output
non_padded_labels = [label for label in example_labels if label != -100]

# The model predicts probabilities for each token *in the input sequence* for the *next* token.
# So, if input has length L, output has shape (L, vocab_size).
# The target for position `i`'s prediction is `input[i+1]`. Our `example_labels` are already this.

# The gpt_model inherently handles sequence processing up to its length.
# We need to ensure that the prediction for position `j` is compared to `labels[j]`

# Make a dummy call to gpt_model for the example sequence
# In a real setup, `gpt_model` would be adapted to handle batches.
# Here, `probabilities` will be (seq_len, vocab_size)

# Ensure d_model, d_ff, num_heads, num_decoder_blocks, new_vocab_size are defined
# from previous cells to make this executable.
probabilities = gpt_model(example_input_ids, embedding_matrix_tinystories, new_vocab_size, d_model, d_ff, num_heads, num_decoder_blocks)

# Calculate loss for this single example
loss = categorical_cross_entropy_loss(predictions=probabilities, targets=np.array(example_labels), ignore_index=-100)

print(f"\nShape of model output probabilities: {probabilities.shape}")
print(f"Calculated Conceptual Loss for first example: {loss:.4f}")
print("\nSince the model is untrained, this loss value will be high.")

Example input_ids (first 10): [3198, 1110, 11, 257, 1310, 2576, 3706, 20037, 1043, 257]...
Example labels (first 10): [3198, 1110, 11, 257, 1310, 2576, 3706, 20037, 1043, 257]...


NameError: name 'gpt_decoder_block' is not defined

### 4. Optimization (Conceptual)

After calculating the loss, the next step in a real training process would be **optimization**:

1.  **Backward Pass (Gradient Calculation):** Compute the gradients of the loss with respect to all trainable parameters (e.g., `embedding_matrix`, weights in `gpt_decoder_block`, `W_out`, `b_out`). This involves backpropagation.
2.  **Parameter Update:** Use an optimizer (like Stochastic Gradient Descent, Adam, etc.) to adjust the parameters in the direction that minimizes the loss.

**Note on NumPy Implementation:** Implementing a full backward pass and an optimizer from scratch in NumPy for a complex model like a Transformer is a highly involved task, typically requiring a custom automatic differentiation engine. This is why deep learning frameworks are essential for practical model training.

For our conceptual model, we will **not be implementing** the backward pass or parameter updates. The model weights (`embedding_matrix_tinystories`, `W_out`, `b_out`, and internal weights of `feed_forward_network`) remain randomly initialized throughout this conceptual build. This is why our text generation remains incoherent.

### Demonstration of the GPT Model

Let's run our full GPT model with a sample input sequence and observe its output probabilities for the next token at each position. We'll use the same input sequence `['the', 'cat', 'chased', 'the']`.

In [ ]:
# Parameters for the GPT model
num_decoder_blocks = 2 # As per the architecture sketch

print(f"GPT Model Parameters:")
print(f"  Vocabulary size: {vocab_size}")
print(f"  Embedding dimension (d_model): {d_model}")
print(f"  Feed Forward inner dimension (d_ff): {d_ff}")
print(f"  Number of attention heads: {num_heads}")
print(f"  Number of decoder blocks: {num_decoder_blocks}")

# Use the sample token IDs from previous tasks
sample_token_ids_model = sample_token_ids # [13, 1, 2, 13] for 'the cat chased the'

print(f"\nInput token IDs: {sample_token_ids_model}")
print(f"Input sequence words: {[idx_to_word[idx] for idx in sample_token_ids_model]}")

# Run the GPT model
output_probabilities = gpt_model(sample_token_ids_model, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks)

print("\nGPT Model Output Probabilities Shape:", output_probabilities.shape)

print("\nOutput Probabilities for each token in the sequence (top 3 predictions):")
for i, token_id in enumerate(sample_token_ids_model):
    print(f"--- Position {i} (Input: '{idx_to_word[token_id]}') ---")
    # Get top 3 predicted tokens and their probabilities
    top_indices = np.argsort(output_probabilities[i])[::-1][:3]
    for idx in top_indices:
        word = idx_to_word[idx]
        prob = output_probabilities[i, idx]
        print(f"  '{word}': {prob:.4f}")

print("\nThis demonstrates the full forward pass of a GPT-like model, predicting the next token at each step.")

GPT Model Parameters:
  Vocabulary size: 15
  Embedding dimension (d_model): 8
  Feed Forward inner dimension (d_ff): 16
  Number of attention heads: 4
  Number of decoder blocks: 2

Input token IDs: [13, 1, 2, 13]
Input sequence words: ['the', 'cat', 'chased', 'the']

GPT Model Output Probabilities Shape: (4, 15)

Output Probabilities for each token in the sequence (top 3 predictions):
--- Position 0 (Input: 'the') ---
  'machine': 0.0672
  'dog': 0.0672
  'mouse': 0.0670
--- Position 1 (Input: 'cat') ---
  'machine': 0.0679
  'loudly': 0.0675
  'chased': 0.0673
--- Position 2 (Input: 'chased') ---
  'neural': 0.0677
  'barked': 0.0671
  'dog': 0.0670
--- Position 3 (Input: 'the') ---
  'machine': 0.0672
  'dog': 0.0672
  'neural': 0.0670

This demonstrates the full forward pass of a GPT-like model, predicting the next token at each step.


## Task 8: Next Token Prediction

This task involves using the trained GPT model to predict the most likely next word given an input sequence. We'll feed the sequence into our `gpt_model` and take the token with the highest probability at the last position as our prediction.

In [ ]:
def predict_next_token(input_token_ids, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, idx_to_word):
    """
    Predicts the next token given an input sequence.
    """
    if not input_token_ids:
        return "", 0.0 # No input, no prediction

    # Run the GPT model to get probabilities
    probabilities = gpt_model(input_token_ids, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks)

    # Get the probabilities for the last token in the sequence (which predicts the next token)
    last_token_probabilities = probabilities[-1]

    # Get the index of the token with the highest probability
    predicted_token_idx = np.argmax(last_token_probabilities)
    predicted_probability = last_token_probabilities[predicted_token_idx]

    return idx_to_word[predicted_token_idx], predicted_probability

print("Defined predict_next_token function.")

Defined predict_next_token function.


### Demonstration of Next Token Prediction

In [ ]:
# Example 1: "The cat chased"
input_phrase_1 = "the cat chased".split()
input_token_ids_1 = [word_to_idx[word] for word in input_phrase_1 if word in word_to_idx]

predicted_word_1, prob_1 = predict_next_token(input_token_ids_1, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, idx_to_word)
print(f"Input: '{' '.join(input_phrase_1)}'\nPrediction: '{predicted_word_1}' (Probability: {prob_1:.4f})\n")

# Example 2: "The cat chased the"
input_phrase_2 = "the cat chased the".split()
input_token_ids_2 = [word_to_idx[word] for word in input_phrase_2 if word in word_to_idx]

predicted_word_2, prob_2 = predict_next_token(input_token_ids_2, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, idx_to_word)
print(f"Input: '{' '.join(input_phrase_2)}'\nPrediction: '{predicted_word_2}' (Probability: {prob_2:.4f})\n")

Input: 'the cat chased'
Prediction: 'is' (Probability: 0.0677)

Input: 'the cat chased the'
Prediction: 'deep' (Probability: 0.0679)



## Task 9: Text Generation (Autoregressive Generation)

Text generation involves iteratively predicting the next token and appending it to the input sequence, then using the extended sequence to predict the next token. This process is repeated until a desired length is reached, or a special token (like an end-of-sentence token, which we don't have in this simplified model) is generated.

### Explain: Autoregressive Generation

Autoregressive generation means that each new token is generated conditioned on all previously generated tokens (and the initial prompt). The model predicts one token at a time, adds it to the sequence, and then uses that updated sequence to predict the next token, creating a chain reaction. This is fundamental to how GPT models produce coherent and contextually relevant text.

In [ ]:
def generate_text(prompt_token_ids, max_length, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, idx_to_word):
    """
    Generates text autoregressively from a given prompt.
    """
    generated_token_ids = list(prompt_token_ids)

    for _ in range(max_length - len(prompt_token_ids)): # Generate tokens until max_length is reached
        if not generated_token_ids: # Handle empty sequence if max_length is less than prompt length
            break

        # Predict the next token based on the current generated sequence
        probabilities = gpt_model(generated_token_ids, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks)

        # Get probabilities for the last token's prediction
        last_token_probabilities = probabilities[-1]

        # Select the token with the highest probability (greedy decoding)
        next_token_idx = np.argmax(last_token_probabilities)
        generated_token_ids.append(next_token_idx)

    # Convert token IDs back to words
    generated_words = [idx_to_word[idx] for idx in generated_token_ids if idx in idx_to_word]
    return ' '.join(generated_words)

print("Defined generate_text function.")

Defined generate_text function.


### Demonstration of Text Generation

In [ ]:
# Example: "Machine learning"
prompt_phrase = "machine learning".split()
prompt_token_ids = [word_to_idx[word] for word in prompt_phrase if word in word_to_idx]

# Generate text of a certain length (e.g., 5 tokens)
generated_text = generate_text(prompt_token_ids, len(prompt_token_ids) + 3, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, idx_to_word)
print(f"Input Prompt: '{' '.join(prompt_phrase)}'\nGeneration: {generated_text}")

# Note: Due to random initialization and small vocabulary, the generations might not be semantically coherent.

Input Prompt: 'machine learning'
Generation: machine learning learning chased cat


# **project**

# AI Story Generator using `roneneldan/TinyStories` Dataset

## Task: Data Preparation (TinyStories)

### Install Libraries

In [ ]:
!pip install datasets transformers numpy

print("Installed required libraries: datasets, transformers, numpy")

Installed required libraries: datasets, transformers, numpy


### Load Dataset

In [ ]:
from datasets import load_dataset

print("Loading the 'roneneldan/TinyStories' dataset...")
dataset = load_dataset("roneneldan/TinyStories")

print("Dataset loaded successfully!")
print(dataset)
print("First example from the training split:")
print(dataset['train'][0])

Loading the 'roneneldan/TinyStories' dataset...
Dataset loaded successfully!
DatasetDict({
    train: Dataset({
        features: ['text'],
        num_rows: 2119719
    })
    validation: Dataset({
        features: ['text'],
        num_rows: 21990
    })
})
First example from the training split:
{'text': 'One day, a little girl named Lily found a needle in her room. She knew it was difficult to play with it because it was sharp. Lily wanted to share the needle with her mom, so she could sew a button on her shirt.\n\nLily went to her mom and said, "Mom, I found this needle. Can you share it with me and sew my shirt?" Her mom smiled and said, "Yes, Lily, we can share the needle and fix your shirt."\n\nTogether, they shared the needle and sewed the button on Lily\'s shirt. It was not difficult for them because they were sharing and helping each other. After they finished, Lily thanked her mom for sharing the needle and fixing her shirt. They both felt happy because they had shared and 

## Task: Tokenization (TinyStories)

We'll use a pre-trained `GPT2Tokenizer` to prepare our text data for the GPT model. This involves:

1.  **Loading the tokenizer:** Instantiating `AutoTokenizer` with 'gpt2'.
2.  **Tokenizing a subset of the dataset:** Applying the tokenizer to a smaller subset of the 'text' column for demonstration due to processing time.
3.  **Handling special tokens:** Adding padding and end-of-sequence tokens as needed for model input.

### Tokenizer Setup

In [ ]:
from transformers import AutoTokenizer
import numpy as np # Ensure numpy is imported

# Load pre-trained tokenizer
tokenizer = AutoTokenizer.from_pretrained("gpt2")

# Add a padding token if it doesn't exist (GPT2 tokenizer often doesn't have one by default)
if tokenizer.pad_token is None:
    tokenizer.add_special_tokens({'pad_token': tokenizer.eos_token}) # Use EOS as PAD for GPT-like models

print(f"Tokenizer loaded: {tokenizer.name_or_path}")
print(f"Vocabulary size: {len(tokenizer)}")
print(f"EOS token ID: {tokenizer.eos_token_id}")
print(f"PAD token ID: {tokenizer.pad_token_id}")

# Define new_vocab_size for TinyStories
new_vocab_size = len(tokenizer)

Tokenizer loaded: gpt2
Vocabulary size: 50257
EOS token ID: 50256
PAD token ID: 50256


### Define Tokenization Function and Apply to Dataset Subset

We'll create a simple function to tokenize each example and then apply it to a subset of the dataset using `map`. We'll also ensure that the sequences are truncated or padded to a maximum length, and include `labels` which are typically the `input_ids` shifted for language modeling tasks.

In [ ]:
max_seq_length = 128 # Define a maximum sequence length

def tokenize_function(examples):
    # The tokenizer automatically handles adding BOS/EOS tokens and padding/truncation.
    # We're also preparing 'labels' for language modeling here.
    tokenized_output = tokenizer(examples["text"],
                                 truncation=True,
                                 max_length=max_seq_length,
                                 padding="max_length") # Pad to max_length

    # For Causal Language Modeling, labels are usually the same as input_ids
    # but with masked padding. We'll set labels to -100 where input_ids are pad_token_id.
    tokenized_output["labels"] = tokenized_output["input_ids"].copy()
    # Replace padding token id in labels with -100, so it's ignored in loss calculation
    tokenized_output["labels"] = [
        [(l if l != tokenizer.pad_token_id else -100) for l in label_seq]
        for label_seq in tokenized_output["labels"]
    ]
    return tokenized_output

# Define the size of the subset to tokenize
subset_size = 30000 # As requested by the user

# Create a subset of the training dataset
print(f"Creating a subset of {subset_size} samples from the 'train' split...")
dataset_subset = dataset['train'].select(range(subset_size))

# Apply the tokenization function to the dataset subset
print(f"Tokenizing the dataset subset with max_seq_length={max_seq_length}...")
tokenized_dataset_subset = dataset_subset.map(tokenize_function, batched=True, remove_columns=["text"])

print("Dataset subset tokenization complete!")
print(tokenized_dataset_subset)
print("First tokenized example:")
print(tokenized_dataset_subset[0])

Creating a subset of 30000 samples from the 'train' split...
Tokenizing the dataset subset with max_seq_length=128...
Dataset subset tokenization complete!
Dataset({
    features: ['input_ids', 'attention_mask', 'labels'],
    num_rows: 30000
})
First tokenized example:
{'input_ids': [3198, 1110, 11, 257, 1310, 2576, 3706, 20037, 1043, 257, 17598, 287, 607, 2119, 13, 1375, 2993, 340, 373, 2408, 284, 711, 351, 340, 780, 340, 373, 7786, 13, 20037, 2227, 284, 2648, 262, 17598, 351, 607, 1995, 11, 523, 673, 714, 34249, 257, 4936, 319, 607, 10147, 13, 198, 198, 43, 813, 1816, 284, 607, 1995, 290, 531, 11, 366, 29252, 11, 314, 1043, 428, 17598, 13, 1680, 345, 2648, 340, 351, 502, 290, 34249, 616, 10147, 1701, 2332, 1995, 13541, 290, 531, 11, 366, 5297, 11, 20037, 11, 356, 460, 2648, 262, 17598, 290, 4259, 534, 10147, 526, 198, 198, 41631, 11, 484, 4888, 262, 17598, 290, 384, 19103, 262, 4936, 319, 20037, 338, 10147, 13, 632, 373, 407, 2408, 329, 606, 780, 484, 547, 7373], 'attention_mask': [

## Task: Embeddings for TinyStories Dataset

Now that we have tokenized the `TinyStories` dataset, we need to create an embedding layer that can represent these tokens as dense vectors. We'll use the vocabulary size from our `gpt2` tokenizer to initialize this embedding matrix. For now, we'll initialize these embeddings randomly, as they would be learned during the training process of a real GPT model.

In [ ]:
# Reuse the embedding_dim defined in the conceptual model (currently 8)
# Make sure d_model is defined or set it here if not
d_model = 8 # This should match the `embedding_dim` used previously

# Initialize a new random embedding matrix for the TinyStories vocabulary
# In a real scenario, these would be learned during training or loaded from pre-trained models.
embedding_matrix_tinystories = np.random.rand(new_vocab_size, d_model)

print(f"New vocabulary size (from GPT2 tokenizer): {new_vocab_size}")
print(f"Embedding dimension (d_model): {d_model}")
print(f"Shape of new embedding matrix: {embedding_matrix_tinystories.shape}")
print("First 5 embeddings (for demonstration):")
for i in range(min(5, new_vocab_size)):
    # Using tokenizer.decode for demonstration, as we don't have idx_to_word for the full vocab
    word = tokenizer.decode([i])
    print(f"Token ID: {i}, Word: '{word.strip()}' -> Embedding: {embedding_matrix_tinystories[i][:5]}...")

New vocabulary size (from GPT2 tokenizer): 50257
Embedding dimension (d_model): 8
Shape of new embedding matrix: (50257, 8)
First 5 embeddings (for demonstration):
Token ID: 0, Word: '!' -> Embedding: [0.28956125 0.92911423 0.1985418  0.70728916 0.65872143]...
Token ID: 1, Word: '"' -> Embedding: [0.08295658 0.35494423 0.44281442 0.62584042 0.76954409]...
Token ID: 2, Word: '#' -> Embedding: [0.83525405 0.86480309 0.78654541 0.68645982 0.76402448]...
Token ID: 3, Word: '$' -> Embedding: [0.40375576 0.41353088 0.5281711  0.40906205 0.83160644]...
Token ID: 4, Word: '%' -> Embedding: [0.63600369 0.99727912 0.79203871 0.27058609 0.42858549]...


## Task: Positional Encoding for TinyStories

Positional encoding is crucial in Transformer models as it injects information about the relative or absolute position of tokens in a sequence. Since our `gpt2` tokenizer has a vocabulary size of 50257 and we're working with sequences up to `max_seq_length=128`, we need a robust positional encoding strategy.

Here, we'll implement the standard sine/cosine positional encoding. This method generates a unique vector for each position, which is then added to the token embeddings before being fed into the decoder blocks.

In [ ]:
import math

def create_sinusoidal_positional_encoding(seq_len, d_model):
    """
    Generates sinusoidal positional encodings.
    seq_len: The maximum sequence length.
    d_model: The dimension of the embeddings.
    """
    position = np.arange(seq_len)[:, np.newaxis]
    div_term = np.exp(np.arange(0, d_model, 2) * -(math.log(10000.0) / d_model))

    pe = np.zeros((seq_len, d_model))
    pe[:, 0::2] = np.sin(position * div_term)
    pe[:, 1::2] = np.cos(position * div_term)

    return pe

# Generate positional encodings for our max_seq_length and d_model
positional_encoding_matrix = create_sinusoidal_positional_encoding(max_seq_length, d_model)

print(f"Positional Encoding Matrix Shape: {positional_encoding_matrix.shape}")
print("First 5 elements of positional encoding for position 0 (first token):")
print(positional_encoding_matrix[0, :5])
print("First 5 elements of positional encoding for position 1 (second token):")
print(positional_encoding_matrix[1, :5])

print("\nThis sinusoidal positional encoding will be added to our token embeddings to provide order information.")

Positional Encoding Matrix Shape: (128, 8)
First 5 elements of positional encoding for position 0 (first token):
[0. 1. 0. 1. 0.]
First 5 elements of positional encoding for position 1 (second token):
[0.84147098 0.54030231 0.09983342 0.99500417 0.00999983]

This sinusoidal positional encoding will be added to our token embeddings to provide order information.


## Task: Consolidate GPT Model and Helper Functions (TinyStories Adapted)

To ensure all necessary functions are within the correct scope and to fix persistent `NameError` issues, we will re-define all core GPT model helper functions and the `gpt_model` itself in a single cell. This `gpt_model` will now utilize the `create_sinusoidal_positional_encoding` function and the `positional_encoding_matrix` generated specifically for the `TinyStories` project context.

In [ ]:
import numpy as np
import math

def softmax(x):
    e_x = np.exp(x - np.max(x, axis=-1, keepdims=True)) # Subtract max for numerical stability
    return e_x / e_x.sum(axis=-1, keepdims=True)

def add_and_normalize(x, sublayer_output, epsilon=1e-6):
    """
    Applies residual connection followed by layer normalization.
    x: The input to the sub-layer (for residual connection).
    sublayer_output: The output of the sub-layer.
    epsilon: Small value to prevent division by zero in normalization.
    """
    # Add residual connection
    added_output = x + sublayer_output

    # Layer Normalization
    mean = np.mean(added_output, axis=-1, keepdims=True)
    variance = np.var(added_output, axis=-1, keepdims=True)
    normalized_output = (added_output - mean) / np.sqrt(variance + epsilon)

    # In a real model, learnable gamma and beta parameters would be applied here.
    # For this conceptual implementation, we omit them.
    return normalized_output

def feed_forward_network(x, d_model, d_ff):
    """
    Applies a simple feed-forward network to the input.
    x: Input tensor.
    d_model: Dimension of input/output.
    d_ff: Dimension of the inner layer of the FFN.
    """
    # Simulate linear transformations (weights and biases are omitted for simplicity)
    # First linear layer (d_model -> d_ff)
    # Assuming random weights for demonstration
    W1 = np.random.rand(d_model, d_ff) * 0.01 # Small weights for stability
    b1 = np.random.rand(d_ff) * 0.01
    linear1 = np.dot(x, W1) + b1

    # ReLU activation
    relu_output = np.maximum(0, linear1)

    # Second linear layer (d_ff -> d_model)
    W2 = np.random.rand(d_ff, d_model) * 0.01
    b2 = np.random.rand(d_model) * 0.01
    output = np.dot(relu_output, W2) + b2

    return output

def gpt_decoder_block(x, d_model, d_ff, num_heads, causal_mask):
    """
    Implements a single GPT Decoder Block, including internal Masked Multi-Head Attention.
    x: The input to the block (e.g., embedded input + positional encoding).
    d_model: The embedding dimension (also the output dimension of the block).
    d_ff: The inner dimension of the Feed Forward Network.
    num_heads: Number of attention heads for Multi-Head Attention.
    causal_mask: The causal mask to apply during attention.
    """

    # 1. Masked Multi-Head Attention
    head_dim = d_model // num_heads
    all_head_outputs = []

    # Simulate linear projections by slicing the embedding dimension for each head
    # In a real model, Q, K, V would be distinct linear transformations of input 'x'
    # For simplicity, we use 'x' as source for Q, K, V and slice it for heads.
    for i in range(num_heads):
        head_start_idx = i * head_dim
        head_end_idx = (i + 1) * head_dim

        Q_head = np.array([vec[head_start_idx:head_end_idx] for vec in x])
        K_head = np.array([vec[head_start_idx:head_end_idx] for vec in x])
        V_head = np.array([vec[head_start_idx:head_end_idx] for vec in x])

        # Calculate raw attention scores for this head
        attention_scores_head = np.matmul(Q_head, K_head.T)
        attention_scores_head = attention_scores_head / np.sqrt(head_dim)

        # Apply causal mask
        masked_attention_scores_head = attention_scores_head + causal_mask

        # Calculate attention weights for this head
        attention_weights_head = softmax(masked_attention_scores_head)

        # Compute output for this head (Attention Weights @ V_head)
        output_head = np.matmul(attention_weights_head, V_head)
        all_head_outputs.append(output_head)

    multi_head_output = np.concatenate(all_head_outputs, axis=-1)

    # 2. Add & Normalize after Masked Multi-Head Attention
    # The input `x` serves as the residual connection for the attention sub-layer.
    norm1_output = add_and_normalize(x, multi_head_output)

    # 3. Feed Forward Network
    ffn_output = feed_forward_network(norm1_output, d_model, d_ff)

    # 4. Add & Normalize after Feed Forward Network
    # `norm1_output` serves as the residual connection for the FFN sub-layer.
    decoder_block_output = add_and_normalize(norm1_output, ffn_output)

    return decoder_block_output

def create_sinusoidal_positional_encoding(seq_len_pe, d_model_pe):
    """
    Generates sinusoidal positional encodings.
    seq_len_pe: The maximum sequence length.
    d_model_pe: The dimension of the embeddings.
    """
    position = np.arange(seq_len_pe)[:, np.newaxis]
    div_term = np.exp(np.arange(0, d_model_pe, 2) * -(math.log(10000.0) / d_model_pe))

    pe = np.zeros((seq_len_pe, d_model_pe))
    pe[:, 0::2] = np.sin(position * div_term)
    pe[:, 1::2] = np.cos(position * div_term)

    return pe

def gpt_model(token_ids_sequence, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks):
    """
    Implements a simplified GPT Model architecture.
    token_ids_sequence: List of token IDs (e.g., [13, 1, 2])
    embedding_matrix: Pre-initialized word embeddings (vocab_size, d_model)
    vocab_size: Size of the vocabulary
    d_model: Dimension of embeddings and decoder block outputs
    d_ff: Inner dimension of the Feed Forward Networks
    num_heads: Number of attention heads
    num_decoder_blocks: Number of stacked decoder blocks
    """
    seq_len = len(token_ids_sequence)

    # 1. Embedding Layer
    embedded_input = np.array([embedding_matrix[idx] for idx in token_ids_sequence])

    # 2. Positional Encoding
    # Use the global max_seq_length (from cell 77c3cec2) to create the positional encoding matrix
    global max_seq_length
    positional_encoding_matrix_local = create_sinusoidal_positional_encoding(max_seq_length, d_model)

    if seq_len > max_seq_length:
        raise ValueError(f"Sequence length ({seq_len}) exceeds max_seq_length ({max_seq_length}) for positional encoding.")

    pos_encoding = positional_encoding_matrix_local[:seq_len, :]
    x = embedded_input + pos_encoding # Add positional information to embeddings

    # Create causal mask for the entire sequence (needed for each decoder block)
    causal_mask_model = np.full((seq_len, seq_len), -np.inf)
    for i in range(seq_len):
        for j in range(i + 1):
            causal_mask_model[i, j] = 0

    # 3. Stack Decoder Blocks
    for block_num in range(num_decoder_blocks):
        x = gpt_decoder_block(x, d_model, d_ff, num_heads, causal_mask_model)

    # 4. Linear Layer (Projection to vocabulary size)
    # Simulate a linear layer: output = x @ W_out + b_out
    # Initialize random weights and biases for demonstration
    W_out = np.random.rand(d_model, vocab_size) * 0.01
    b_out = np.random.rand(vocab_size) * 0.01
    logits = np.dot(x, W_out) + b_out

    # 5. Softmax to get probabilities
    probabilities = softmax(logits)

    return probabilities

print("Defined gpt_model function and its dependencies (softmax, add_and_normalize, feed_forward_network, gpt_decoder_block, create_sinusoidal_positional_encoding) in a consolidated manner.")

Defined gpt_model function and its dependencies (softmax, add_and_normalize, feed_forward_network, gpt_decoder_block, create_sinusoidal_positional_encoding) in a consolidated manner.


### Define Global Parameters for TinyStories GPT Model

In [ ]:
# Define global parameters for the GPT model in the TinyStories context
d_ff = 2 * d_model # Inner dimension of the Feed Forward Network
num_decoder_blocks = 2 # Number of stacked decoder blocks
num_heads = 4 # Number of attention heads

print(f"Global Parameters for TinyStories GPT model:")
print(f"  d_model: {d_model}")
print(f"  d_ff: {d_ff}")
print(f"  num_decoder_blocks: {num_decoder_blocks}")
print(f"  num_heads: {num_heads}")

Global Parameters for TinyStories GPT model:
  d_model: 8
  d_ff: 16
  num_decoder_blocks: 2
  num_heads: 4


## Task: Text Generation Functions (TinyStories Adapted)

Now that our `gpt_model` is correctly defined and uses the `TinyStories` specific components, we will redefine the `predict_next_token_tinystories` and `generate_text_tinystories` functions to ensure they use the newly updated `gpt_model` in their scope.

In [ ]:
def predict_next_token_tinystories(input_text, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, tokenizer):
    """
    Predicts the next token given an input text using the TinyStories-adapted model.
    """
    # Tokenize the input text using the GPT2 tokenizer
    input_token_ids = tokenizer.encode(input_text, return_tensors='np')[0].tolist()

    if not input_token_ids:
        return "", 0.0 # No input, no prediction

    # Run the GPT model to get probabilities
    probabilities = gpt_model(input_token_ids, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks)

    # Get the probabilities for the last token in the sequence (which predicts the next token)
    last_token_probabilities = probabilities[-1]

    # Get the index of the token with the highest probability
    predicted_token_idx = np.argmax(last_token_probabilities)
    predicted_probability = last_token_probabilities[predicted_token_idx]

    # Decode the predicted token back to a word
    predicted_word = tokenizer.decode([predicted_token_idx])

    return predicted_word, predicted_probability

def generate_text_tinystories(prompt_text, max_length, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, tokenizer):
    """
    Generates text autoregressively from a given prompt using the TinyStories-adapted model.
    """
    # Tokenize the initial prompt
    generated_token_ids = tokenizer.encode(prompt_text, return_tensors='np')[0].tolist()

    for _ in range(max_length - len(generated_token_ids)): # Generate tokens until max_length is reached
        if not generated_token_ids: # Handle empty sequence if max_length is less than prompt length
            break

        # Predict the next token based on the current generated sequence
        probabilities = gpt_model(generated_token_ids, embedding_matrix, vocab_size, d_model, d_ff, num_heads, num_decoder_blocks)

        # Get probabilities for the last token's prediction
        last_token_probabilities = probabilities[-1]

        # Select the token with the highest probability (greedy decoding)
        next_token_idx = np.argmax(last_token_probabilities)
        generated_token_ids.append(next_token_idx)

    # Convert token IDs back to words
    generated_text = tokenizer.decode(generated_token_ids)
    return generated_text

print("Redefined predict_next_token_tinystories and generate_text_tinystories functions to ensure correct gpt_model usage.")

Redefined predict_next_token_tinystories and generate_text_tinystories functions to ensure correct gpt_model usage.


### Demonstration: Next Token Prediction for TinyStories

In [ ]:
# Example 1: "The little cat"
input_text_1 = "The little cat"

predicted_word_1, prob_1 = predict_next_token_tinystories(input_text_1, embedding_matrix_tinystories, new_vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, tokenizer)
print(f"Input: '{input_text_1}'\nPrediction: '{predicted_word_1}' (Probability: {prob_1:.4f})\n")

# Example 2: "Once upon a time, a"
input_text_2 = "Once upon a time, a"

predicted_word_2, prob_2 = predict_next_token_tinystories(input_text_2, embedding_matrix_tinystories, new_vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, tokenizer)
print(f"Input: '{input_text_2}'\nPrediction: '{predicted_word_2}' (Probability: {prob_2:.4f})\n")

print("Note: Predictions will still be random as the model is untrained.")

Input: 'The little cat'
Prediction: ' Drum' (Probability: 0.0000)

Input: 'Once upon a time, a'
Prediction: 'aun' (Probability: 0.0000)

Note: Predictions will still be random as the model is untrained.


### Demonstration: Text Generation for TinyStories

In [ ]:
# Example: "The quick brown fox"
prompt_text_gen = "The quick brown fox"
max_gen_length = len(tokenizer.encode(prompt_text_gen)) + 10 # Generate 10 new tokens

generated_text = generate_text_tinystories(prompt_text_gen, max_gen_length, embedding_matrix_tinystories, new_vocab_size, d_model, d_ff, num_heads, num_decoder_blocks, tokenizer)
print(f"Input Prompt: '{prompt_text_gen}'\nGeneration: {generated_text}")

print("\nNote: Text generation will be semantically incoherent because the model's weights are randomly initialized.")

Input Prompt: 'The quick brown fox'
Generation: The quick brown fox bestbysGuy Cats Ring charred sorting Lange implementing Anaheim

Note: Text generation will be semantically incoherent because the model's weights are randomly initialized.


## Task: Training the GPT Model (Conceptual for TinyStories)

Training a GPT model involves iteratively adjusting its weights so that it can accurately predict the next token in a sequence. For our conceptual NumPy model, we'll outline the key steps involved, rather than implementing a full, optimized training loop (which is typically handled by deep learning frameworks like PyTorch or TensorFlow).

The main components of training are:
1.  **Data Preparation:** Organizing the tokenized data into input sequences and corresponding target labels.
2.  **Loss Function:** A metric to quantify how well the model's predictions align with the actual next tokens.
3.  **Optimization (Conceptual):** Adjusting model weights based on the calculated loss (we'll only describe this conceptually).

Let's first prepare our data for this conceptual training.

### 1. Data Preparation for Training Batches

Our `tokenized_dataset_subset` contains `input_ids`, `attention_mask`, and `labels`. For training, we primarily need the `input_ids` as input to our `gpt_model` and `labels` as the ground truth for loss calculation. We'll extract these from our subset.

In [ ]:
# Extract input_ids and labels from the tokenized dataset subset
training_input_ids = tokenized_dataset_subset['input_ids']
training_labels = tokenized_dataset_subset['labels']

print(f"Number of training samples: {len(training_input_ids)}")
print(f"Shape of first input_ids sequence: {np.array(training_input_ids[0]).shape}")
print(f"Shape of first labels sequence: {np.array(training_labels[0]).shape}")

# We'll use a small batch size for conceptual demonstration
batch_size = 4

print(f"\nConceptual Batch Size for Training: {batch_size}")
print("In a real training scenario, these would be fed to the model in batches.")

Number of training samples: 30000
Shape of first input_ids sequence: (128,)
Shape of first labels sequence: (128,)

Conceptual Batch Size for Training: 4
In a real training scenario, these would be fed to the model in batches.


### 2. Loss Function (Categorical Cross-Entropy)

For language modeling, the goal is to predict the next token. This is a multi-class classification problem where each token in the vocabulary is a possible class. **Categorical Cross-Entropy Loss** is typically used to measure the difference between the model's predicted probability distribution over the vocabulary and the true one-hot encoded distribution of the next token.

Since our `labels` are masked with -100 for padding, we need to ensure our loss function ignores these values.

In [ ]:
def categorical_cross_entropy_loss(predictions, targets, ignore_index=-100):
    """
    Calculates categorical cross-entropy loss, ignoring a specified index.

    predictions: (seq_len, vocab_size) array of predicted probabilities.
    targets: (seq_len,) array of true token indices.
    ignore_index: The index to ignore in the target sequence (e.g., padding).
    """
    # Ensure targets are valid indices
    valid_indices = targets != ignore_index
    if not np.any(valid_indices):
        return 0.0 # No valid tokens to compute loss for

    # Filter predictions and targets for valid tokens
    valid_predictions = predictions[valid_indices]
    valid_targets = targets[valid_indices]

    # Clip probabilities to avoid log(0)
    epsilon = 1e-9
    valid_predictions = np.clip(valid_predictions, epsilon, 1. - epsilon)

    # Convert valid_targets to one-hot encoding for comparison with predictions
    # Note: This is simplified; in full model, targets would be shifted for next-token pred.
    # For simplicity here, we assume predictions are for each position's target.
    one_hot_targets = np.zeros_like(valid_predictions)
    one_hot_targets[np.arange(len(valid_targets)), valid_targets] = 1

    loss = -np.sum(one_hot_targets * np.log(valid_predictions)) / len(valid_targets)
    return loss

print("Defined categorical_cross_entropy_loss function.")

Defined categorical_cross_entropy_loss function.


### 3. Conceptual Forward Pass for Training & Loss Calculation

In a training loop, for each batch:
1.  **Input Preparation:** Get a batch of `input_ids`.
2.  **Model Forward Pass:** Pass the `input_ids` through the `gpt_model` to get `output_probabilities` (shape `(batch_seq_len, vocab_size)` for each sequence).
3.  **Target Preparation:** Get the corresponding `labels` for the batch. These `labels` are typically shifted versions of the `input_ids` (e.g., if input is `[t1, t2, t3]`, target for prediction at `t1` is `t2`, at `t2` is `t3`, etc.). Our `tokenized_dataset_subset` already provides `labels` in this format with padding ignored.
4.  **Loss Calculation:** Compute the `categorical_cross_entropy_loss` between the `output_probabilities` and the `labels`.

In [ ]:
# Conceptual demonstration of loss calculation for a single sequence

# Take the first tokenized example from our subset
example_input_ids_tinystories = training_input_ids[0]
example_labels_tinystories = training_labels[0]

print(f"Example input_ids (first 10): {example_input_ids_tinystories[:10]}...")
print(f"Example labels (first 10): {example_labels_tinystories[:10]}...")

# Perform a forward pass with the untrained model
# In a real setup, `gpt_model` would be adapted to handle batches.
# Here, `probabilities` will be (seq_len, vocab_size)
probabilities_tinystories = gpt_model(example_input_ids_tinystories, embedding_matrix_tinystories, new_vocab_size, d_model, d_ff, num_heads, num_decoder_blocks)

# Calculate loss for this single example
loss_tinystories = categorical_cross_entropy_loss(predictions=probabilities_tinystories, targets=np.array(example_labels_tinystories), ignore_index=-100)

print(f"\nShape of model output probabilities: {probabilities_tinystories.shape}")
print(f"Calculated Conceptual Loss for first example: {loss_tinystories:.4f}")
print("\nSince the model is untrained, this loss value will be high.")

Example input_ids (first 10): [3198, 1110, 11, 257, 1310, 2576, 3706, 20037, 1043, 257]...
Example labels (first 10): [3198, 1110, 11, 257, 1310, 2576, 3706, 20037, 1043, 257]...

Shape of model output probabilities: (128, 50257)
Calculated Conceptual Loss for first example: 10.8259

Since the model is untrained, this loss value will be high.


### 4. Optimization (Conceptual)

After calculating the loss, the next step in a real training process would be **optimization**:

1.  **Backward Pass (Gradient Calculation):** Compute the gradients of the loss with respect to all trainable parameters (e.g., `embedding_matrix_tinystories`, weights in `gpt_decoder_block`, `W_out`, `b_out`). This involves backpropagation.
2.  **Parameter Update:** Use an optimizer (like Stochastic Gradient Descent, Adam, etc.) to adjust the parameters in the direction that minimizes the loss.

**Note on NumPy Implementation:** Implementing a full backward pass and an optimizer from scratch in NumPy for a complex model like a Transformer is a highly involved task, typically requiring a custom automatic differentiation engine. This is why deep learning frameworks are essential for practical model training.

For our conceptual model, we will **not be implementing** the backward pass or parameter updates. The model weights (`embedding_matrix_tinystories`, `W_out`, `b_out`, and internal weights of `feed_forward_network`) remain randomly initialized throughout this conceptual build. This is why our text generation remains incoherent.

## Task: Evaluation (Conceptual for TinyStories)

Evaluating a GPT model typically involves metrics that assess the quality of generated text or the model's ability to predict the next token accurately. Since our model is untrained and its weights are random, a quantitative evaluation (like perplexity or BLEU score) would yield poor results and not be informative for this conceptual demonstration. Instead, we'll conceptually outline what evaluation would entail.

### Conceptual Evaluation Metrics:

*   **Perplexity:** A common metric for language models. It measures how well a probability distribution predicts a sample. Lower perplexity indicates a better model. For an untrained model, perplexity would be extremely high.
*   **BLEU (Bilingual Evaluation Understudy) Score:** Often used for machine translation, but can be adapted for text generation by comparing generated text to reference texts. Since we don't have reference generations for arbitrary prompts, and our model generates incoherent text, this wouldn't be meaningful here.
*   **Human Evaluation:** The most robust way to evaluate generated stories would be to have humans read and rate them based on fluency, coherence, creativity, and adherence to the prompt. This is subjective and beyond the scope of this NumPy implementation.

### Current State of Evaluation for Our Conceptual Model:

Given our randomly initialized weights and lack of training, the primary 'evaluation' we can perform is to observe the output of our `predict_next_token_tinystories` and `generate_text_tinystories` functions. As demonstrated, the predictions and generated text are semantically incoherent, which is the expected behavior of an untrained model. This confirms the structural integrity of the model (it runs without errors), but highlights the absolute necessity of a robust training process for achieving meaningful outputs.